docs/architecture/adr-054-semver-compatible-modules.md

```diff
@@ -0,0 +1,493 @@
+# ADR 054: Semver Compatible SDK Modules
+
+## Changelog
+
+* 2022-04-27: First draft
+
+## Status
+
+PROPOSED
+
+## Abstract
+
+In order to move the Cosmos SDK to a system of decoupled semantically versioned
+modules which can be composed in different combinations (ex. staking v3 with
+bank v1 and distribution v2), we need to reassess how we organize the API surface
```

Comment:
Is this actually a goal? What specific problems does it solve?

Reply:
Yes. People often want to adopt certain upgrades but not others. It has been one of our major pain points. There are valid forks of things like staking that have a common feature set and also some distinct behaviors like liquid staking for example.

Reply:
If you want to allow end users to compose different versions of modules into their binary, then it seems like the obvious play would be to move modules out of the cosmos-sdk monorepo and into their own distinct repos, right?

Reply:
Currently we're planning to stick with a monorepo. Multiple repos would make human coordination much more difficult IMHO

Reply:
I also agree we should be sticking with monorepo

Reply:
> It doesn't matter if the repo is mono repo or not. Go modules solve that. For doing cross version testing you would need to setup new modules anyway (in the same repo or in other repo)

Go modules support multiple modules in a single repository out of necessity, it's not a blessed path.

> For all but power users, you probably want to adopt the usual convention that one repo = one module. It's important for long-term evolution of code storage options that a repo can contain multiple modules, but it's almost certainly not something you want to do by default.

https://github.com/golang/go/wiki/Modules#should-i-have-multiple-modules-in-a-single-repository

Reply:
Multi-module repos unfortunately result in _more_ complexity w.r.t. human coordination, versus separate  repos.